proposals/xxxx-package-manager-target-spec.md

```diff
@@ -0,0 +1,187 @@
+
+## Detailed Design
+
+### Convention based layouts
+
+We propose to introduce a layout property to `Package` class, which is an enum `PackageLayout`:
+
+```swift
+enum PackageLayout {
+    /// Use the convention system to infer the targets in the package.
+    case automatic(sources: PackageSourcesDirectory)
+
+    /// Turn off package manager's convention system and manually define all targets.
+    case custom
+}
+
+enum PackageSourcesDirectory {
+    /// Automatically infer the sources directory in order: Sources, Src, src, package root.
+    case automatic
+
+    /// The sources directory is package root. All valid sources files will be included in target except the manifest file, `Package.swift`.
+    case packageRoot
+
+    /// Assume the sources are present at this directory, this must be a relative path from the package root.
+    case custom(path: String)
+}
+```
+
+The `automatic` package layout would work as follows: 
+* Use existing conventions to infer the targets in the package.
+* The inferred targets still have limited customizable properties like `dependencies`.
+
+Consider the following layout and manifest:
+
+```
+├── MySources
+│   ├── Bar
+│   ├── Baz
+│   └── Foo
+└── Tests
+    └── FooTests
+```
+
+```swift
+let package = Package(
+    name: "MyPackage",
+    layout: .automatic(sources: .custom(path: "MySources")),
+    targets: [
+        Target(
+            name: "Foo", 
+            dependencies: ["Bar", "Baz"]
+        ),
+    ]
+)
+```
+
+Using the existing conventions four targets will be created:
+
+* `Bar`, `Baz`: Swift targets
+* `Foo`: Swift target which depends on Bar and Baz
+* `FooTests`: Swift test target which depends on `Foo`
+
+### Custom Layouts and Targets
+
+We propose to add the following structures to `PackageDescription` module:
+
+```swift
+/// Defines a path for a group of files.
+public enum PathSpec {
+
+    /// Include all the files in this directory. Setting recursive to true will also (recusively) search the subdirectories.
+    case directory(String, recursive: Bool)
+
+    /// Include the file at this path.
+    case file(String)
+}
+
+/// Defines the settings of a target.
+public struct TargetSettings {
+
+    /// The type of target.
+    public enum `Type` {
+
+        /// Swift target.
+        case swift
+
+        /// C target which can have headers and a modulemap.
+        case c(headers: [PathSpec], modulemap: String)
+
+        /// System module target. The pkgConfig will help package manager search for additional compiler and linker flags associated with the system package.
+        case systemModule(pkgConfig: String)
+    }
+
+    /// The type of target.
+    public let type: Type
+    
+    /// If the target is a test target.
+    public let isTest: Bool
+    
+    /// If defined, makes the target executable.
+    public let executable: Bool
```

Comment:
I agree, I think we should have it for all targets, but Swift will still require us to follow the main.swift convention because swiftc itself also requires knowing which file is the main executable file (which one should be allowed to have general purpose top-level code). So it should be an error to declare an executable Swift module but not have that file.